### PR TITLE
Add multi-selection for `SpriteFramesEditor`

### LINKED ISSUE #1696: Allow Multi-Selection of Frames in SpriteFrames Animations
**Describe the project you are working on:**
I'm currently using AnimatedSprites, which use SpriteFrames, to add animations to the user interface of our game project.

**Describe the problem or limitation you are having in your project:**
I have been modifying the frames of our animations in SpriteFrames, by swapping out old sprites and implementing new ones, as well as moving frames around. While doing this I noticed both moving and deleting can only be done on an individual frame basis, which can make removing or moving sprites in large animations a real chore.

image
_only one frame can be selected at a time_

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
I would love to see multi-selection of the frames inside the Animations, by using CTRL+LMB and SHIFT+LMB respectively for single selection (current selected frames + clicked frame)  or range selection (current selected frames to clicked frame). This is how most listed items and UI elements work throughout the engine, such as the Filesystem, Scene Tree, Scene Viewer.

This would allow the user to select all frames, and then press the 'Delete' button to remove all of them, or select a few frames, and press the 'Move' buttons to move the selected range of frames around. This would greatly speed up large changes within SpriteFrames Animations, and be friendlier on the Index Fingers of all Godot Devs.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**

AnimatedSpriteMultiSelectionMockup
I think it should be pretty self-explanatory, but here is a small slideshow of the proposed feature, as I see it. Keep in mind the last frame, where the animation frames get deleted would obviously not look like that (empty frames), but it best showed the 'deletion' process.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
It cannot be worked around through script, but it can be worked around with manual labor, by selecting each individual frame and moving/deleting it one by one. Not a great solution, but it does work. Technically.

**Is there a reason why this should be core and not an add-on in the asset library?:**
This would be an enhancement to the UX of an already core part of the game, and one that would be in line with other selection methods inside the editor. This likely shouldn't be left as an add-on.

### PR DESCRIPTION
**Pull request by Hobitus**
Enabled multi-selection in the sprite_frame_editor
This was proposal [#1696](https://github.com/godotengine/godot-proposals/issues/1696)

You can now use multi-selections with CTR to add or remove a frame. Shift to selection until selection.
With a multi-selection you can:

**Copy and Paste multiple-frame**
Since you can't put arrays of frame in the Editor clipboard, I've remove it and saved the data locally. 

**Move left or right selected frame**
For example, if you have 1 and 3 selected and press move right, you will have 2,1,4,3.

**Set frame duration to all selected frame.**
 Takes the first frame for the duration value.

**Add empty frame, after or before the selection**

**Delete multiple frame at the same time**

image

I didn't see a way to add test for this class. Anything I can do?

<i>Bugsquad edit:</i>
- Closes https://github.com/godotengine/godot-proposals/issues/1696

### CODE REVIEW

**Comment by Chaosus** [comment]
Static check failed, you should apply the clang-format tool to format it properly. After making the change, don't forget to squash commits (see https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html#modifying-a-pull-request).

**Comment by Hobitus** [comment]
I feel like the move left and move right command are over complicated. It moves all the selected animation (with gap in selection), which is great, but is it really useful?
For simplicity we could pick the first or last item in the multi-selection and move that one only.

**Comment by Chaosus** [comment]
> How do I apply the clang-format tool?

https://docs.godotengine.org/en/stable/contributing/development/code_style_guidelines.html#using-clang-format-locally

File: editor/plugins/sprite_frames_editor_plugin.cpp [hunk 1]
```diff
@@ -1261,12 +1320,31 @@ void SpriteFramesEditor::_update_library_impl() {
...
 
+	for (int index = 0; index < selections.size(); index++) {
+		int sel = selections[index];
+		if (sel == -1) {
+			selections.remove_at(index);
+			index--;
+		}
+		if (sel >= anim_frame_count) {
+			selections.set(index, anim_frame_count - 1);
+			// since selection are ordered, if we get a frame that is outside of the range
```
Comment (dalexeev) [inline editor/plugins/sprite_frames_editor_plugin.cpp:1336]:
```suggestion
			// Since selection is ordered, if we get a frame that is outside of the range
```

File: editor/plugins/sprite_frames_editor_plugin.h [hunk 2]
```diff
@@ -115,7 +115,8 @@ class SpriteFramesEditor : public HSplitContainer {
 	SpinBox *frame_duration = nullptr;
 	ItemList *frame_list = nullptr;
 	bool loading_scene;
-	int sel;
+	Vector<int> selections;
```
Comment (dalexeev) [inline editor/plugins/sprite_frames_editor_plugin.h:118]:
```suggestion
	Vector<int> selection;
```

or

```suggestion
	Vector<int> selected_frames;
```

File: editor/plugins/sprite_frames_editor_plugin.cpp [hunk 3]
```diff
@@ -1112,15 +1162,18 @@ void SpriteFramesEditor::_frame_list_gui_input(const Ref<InputEvent> &p_event) {
...
 
-void SpriteFramesEditor::_frame_list_item_selected(int p_index) {
+void SpriteFramesEditor::_frame_list_item_selected(int p_index, bool p_selected) {
 	if (updating) {
 		return;
 	}
 
-	sel = p_index;
+	selections = frame_list->get_selected_items();
+	if (selections.is_empty() || p_selected == false) {
```
Comment (AThousandShips) [inline editor/plugins/sprite_frames_editor_plugin.cpp:1171]:
```suggestion
	if (selections.is_empty() || !p_selected) {
```

File: editor/plugins/sprite_frames_editor_plugin.h (hunk 2 above)
Comment (Hobitus) [reply editor/plugins/sprite_frames_editor_plugin.h:118]: There is already a frames_selected variable, so I'll go with selection

File: editor/plugins/sprite_frames_editor_plugin.h [hunk 4]
```diff
@@ -115,7 +115,8 @@ class SpriteFramesEditor : public HSplitContainer {
...
+	Vector<Ref<EditorSpriteFramesFrame>> frame_clipboard;
```
[the remark this answers is not in the document]
Comment (Hobitus) [reply editor/plugins/sprite_frames_editor_plugin.h:119]: I agree that I should use the resource_clipboard instead of my frame_clipboard. We can only set resource in a resource_clipboard, I'll have to go with the EditorSpriteFramesArray suggestion.

In the code I see only 4 place where we set_resource_clipboard. In the inspector dock to copy the current resource. could be anything, in the editor resource picker(can also be anything), and in the animation_library_editor (it saves an anim resource).

File: editor/plugins/sprite_frames_editor_plugin.cpp [hunk 5]
```diff
@@ -674,30 +674,29 @@ void SpriteFramesEditor::_load_pressed() {
...
+		return;
+	}
+
 	Ref<Texture2D> texture;
 	float duration = 1.0;
 
-	Ref<EditorSpriteFramesFrame> frame = EditorSettings::get_singleton()->get_resource_clipboard();
-	if (frame.is_valid()) {
+	EditorUndoRedoManager *undo_redo = EditorUndoRedoManager::get_singleton();
+	undo_redo->create_action(TTR("Paste Frame"), UndoRedo::MERGE_DISABLE, EditorNode::get_singleton()->get_edited_scene());
```
[the remark this answers is not in the document]
Comment (Hobitus) [reply editor/plugins/sprite_frames_editor_plugin.cpp]: Added the (s) here and for the Copy and Paste command too.

File: editor/plugins/sprite_frames_editor_plugin.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (Hobitus) [reply editor/plugins/sprite_frames_editor_plugin.h]: Rename the EditorSpriteFramesFrame to EditorSpriteFrame (I feel it makes more sense), and created a EditorSpriteFrameArray, that inherits resources to put it in the EditorSettings clipboard.

File: editor/plugins/sprite_frames_editor_plugin.cpp (hunk 1 above)
Comment (Hobitus) [reply editor/plugins/sprite_frames_editor_plugin.cpp]: done

File: editor/plugins/sprite_frames_editor_plugin.cpp (hunk 3 above)
Comment (Hobitus) [reply editor/plugins/sprite_frames_editor_plugin.cpp]: done

File: editor/plugins/sprite_frames_editor_plugin.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (Hobitus) [reply editor/plugins/sprite_frames_editor_plugin.h]: I've also put back the check if we have a Texture in the clipboard, and create a frame from it.

File: editor/plugins/sprite_frames_editor_plugin.h [hunk 6]
```diff
@@ -49,14 +49,27 @@
...
 
 public:
 	Ref<Texture2D> texture;
 	float duration;
 };
 
+class EditorSpriteFrameArray : public Resource {
+	GDCLASS(EditorSpriteFrameArray, Resource);
+
+	Vector<Ref<EditorSpriteFrame>> frames;
```
Comment (dalexeev) [inline editor/plugins/sprite_frames_editor_plugin.h:63]: Sorry, I worded it poorly. There is no point in having both `EditorSpriteFrame` and `EditorSpriteFrameArray` since they are not exposed. I meant something like:

```c++
class EditorSpriteFrameArray : public Resource {
    GDCLASS(EditorSpriteFrameArray, Resource);

public:
    struct Frame {
        Ref<Texture2D> texture;
        float duration;
    }
    Vector<Frame> frames;
};
```

Comment (Hobitus) [reply editor/plugins/sprite_frames_editor_plugin.h]: The main class is SpriteFramesEditor, so I didn't use EditorSpriteFrames, would be confusing.
I picked ClipboardSpriteFrames instead.

**Comment by KoBeWi** [comment]
Deleting frames should also remove selection, otherwise it looks confusing.

File: editor/plugins/sprite_frames_editor_plugin.cpp [hunk 7]
```diff
@@ -674,30 +674,58 @@ void SpriteFramesEditor::_load_pressed() {
...
+		return;
+	}
+
 	Ref<Texture2D> texture;
 	float duration = 1.0;
 
-	Ref<EditorSpriteFramesFrame> frame = EditorSettings::get_singleton()->get_resource_clipboard();
-	if (frame.is_valid()) {
+	EditorUndoRedoManager *undo_redo = EditorUndoRedoManager::get_singleton();
+	undo_redo->create_action(TTR("Paste Frame(s)"), UndoRedo::MERGE_DISABLE, EditorNode::get_singleton()->get_edited_scene());
```
Comment (KoBeWi) [inline editor/plugins/sprite_frames_editor_plugin.cpp:700]:
```suggestion
	undo_redo->create_action(TTR("Paste Frame(s)"), UndoRedo::MERGE_DISABLE, frames.ptr());
```
If frames are external resource, they would be wrongly associated with the opened scene.

File: editor/plugins/sprite_frames_editor_plugin.cpp [hunk 8]
```diff
@@ -674,30 +674,58 @@ void SpriteFramesEditor::_load_pressed() {
...
+}
+
+void SpriteFramesEditor::_paste_texture(const Ref<Texture2D> &p_texture) {
+	float duration = 1.0;
+
 	EditorUndoRedoManager *undo_redo = EditorUndoRedoManager::get_singleton();
-	undo_redo->create_action(TTR("Paste Frame"), UndoRedo::MERGE_DISABLE, EditorNode::get_singleton()->get_edited_scene());
-	undo_redo->add_do_method(frames.ptr(), "add_frame", edited_anim, texture, duration);
-	undo_redo->add_undo_method(frames.ptr(), "remove_frame", edited_anim, frames->get_frame_count(edited_anim));
+	undo_redo->create_action(TTR("Paste Texture"), UndoRedo::MERGE_DISABLE, EditorNode::get_singleton()->get_edited_scene());
```
Comment (KoBeWi) [inline editor/plugins/sprite_frames_editor_plugin.cpp:722]:
```suggestion
	undo_redo->create_action(TTR("Paste Texture"), UndoRedo::MERGE_DISABLE, frames.ptr());
```
Same.
I see originally it was different, but that was wrong.

File: editor/plugins/sprite_frames_editor_plugin.cpp [hunk 9]
```diff
@@ -706,31 +734,41 @@ void SpriteFramesEditor::_paste_pressed() {
...
 	}
 
-	Ref<EditorSpriteFramesFrame> frame = memnew(EditorSpriteFramesFrame);
-	frame->texture = texture;
-	frame->duration = frames->get_frame_duration(edited_anim, frame_list->get_current());
+	Ref<ClipboardSpriteFrames> clipboard_frames = memnew(ClipboardSpriteFrames);
 
-	EditorSettings::get_singleton()->set_resource_clipboard(frame);
+	for (int index = 0; index < selected_items.size(); index++) {
+		int frame_index = selected_items[index];
```
Comment (KoBeWi) [inline editor/plugins/sprite_frames_editor_plugin.cpp:746]:
```suggestion
	for (const int &frame_index : selected_items) {
```

File: editor/plugins/sprite_frames_editor_plugin.cpp [hunk 10]
```diff
@@ -1129,18 +1213,24 @@ void SpriteFramesEditor::_frame_duration_changed(double p_value) {
...
-	Ref<Texture2D> texture = frames->get_frame_texture(edited_anim, index);
-	float old_duration = frames->get_frame_duration(edited_anim, index);
-
 	EditorUndoRedoManager *undo_redo = EditorUndoRedoManager::get_singleton();
 	undo_redo->create_action(TTR("Set Frame Duration"), UndoRedo::MERGE_ENDS, EditorNode::get_singleton()->get_edited_scene());
-	undo_redo->add_do_method(frames.ptr(), "set_frame", edited_anim, index, texture, p_value);
-	undo_redo->add_undo_method(frames.ptr(), "set_frame", edited_anim, index, texture, old_duration);
+
+	for (int selected_index = 0; selected_index < nb_selected_items; selected_index++) {
+		int index = selection[selected_index];
```
Comment (KoBeWi) [inline editor/plugins/sprite_frames_editor_plugin.cpp:1225]:
```suggestion
	for (const int &index  : selection) {
		int index = selection[selected_index];
```

File: editor/plugins/sprite_frames_editor_plugin.cpp [hunk 11]
```diff
@@ -1129,18 +1213,24 @@ void SpriteFramesEditor::_frame_duration_changed(double p_value) {
 		return;
 	}
 
-	int index = sel;
-	if (index < 0) {
+	int nb_selected_items = selection.size();
+	if (nb_selected_items == 0) {
```
Comment (KoBeWi) [inline editor/plugins/sprite_frames_editor_plugin.cpp:1217]:
```suggestion
	if (selection.is_empty()) {
```

**Comment by Hobitus** [comment]
@KoBeWi 
> Deleting frames should also remove selection, otherwise it looks confusing.

From what I tested with the current build of godot (4.2.1). it works like what you see with my PR.
-Pasting new frame keeps the old selection. 
-undoing doesn't change selection. It keeps what you currently have.
-Deleting frame keep the current selection ( 3 frames with the 1 selected, press delete) and its still the 1 that is selected.

You want me to change the current behavior?

**Comment by KoBeWi** [comment]
Looks like the current behavior is not that great :/
You could leave it as is then, it can be improved in a later PR.

File: editor/plugins/sprite_frames_editor_plugin.cpp [hunk 12]
```diff
@@ -674,30 +674,58 @@ void SpriteFramesEditor::_load_pressed() {
 void SpriteFramesEditor::_paste_pressed() {
 	ERR_FAIL_COND(!frames->has_animation(edited_anim));
 
+	Ref<ClipboardSpriteFrames> clipboard_frames = EditorSettings::get_singleton()->get_resource_clipboard();
+
```
[the remark this answers is not in the document]
Comment (Hobitus) [reply editor/plugins/sprite_frames_editor_plugin.cpp]: removed the space

File: editor/plugins/sprite_frames_editor_plugin.cpp (hunk 7 above)
Comment (Hobitus) [reply editor/plugins/sprite_frames_editor_plugin.cpp]: There was many create action in the files. changed them all to use frames.ptr()

File: editor/plugins/sprite_frames_editor_plugin.cpp [hunk 13]
```diff
@@ -674,30 +674,58 @@ void SpriteFramesEditor::_load_pressed() {
...
-	if (frame.is_valid()) {
+	EditorUndoRedoManager *undo_redo = EditorUndoRedoManager::get_singleton();
+	undo_redo->create_action(TTR("Paste Frame(s)"), UndoRedo::MERGE_DISABLE, EditorNode::get_singleton()->get_edited_scene());
+
+	int undo_index = frames->get_frame_count(edited_anim);
+
+	for (int index = 0; index < p_clipboard_frames->frames.size(); index++) {
+		const ClipboardSpriteFrames::Frame *frame = &p_clipboard_frames->frames[index];
 		texture = frame->texture;
 		duration = frame->duration;
```
[the remark this answers is not in the document]
Comment (Hobitus) [reply editor/plugins/sprite_frames_editor_plugin.cpp]: done

File: editor/plugins/sprite_frames_editor_plugin.cpp (hunk 9 above)
Comment (Hobitus) [reply editor/plugins/sprite_frames_editor_plugin.cpp]: done

File: editor/plugins/sprite_frames_editor_plugin.cpp (hunk 10 above)
Comment (Hobitus) [reply editor/plugins/sprite_frames_editor_plugin.cpp]: done

File: editor/plugins/sprite_frames_editor_plugin.cpp (hunk 11 above)
Comment (Hobitus) [reply editor/plugins/sprite_frames_editor_plugin.cpp]: done

**Comment by Hobitus** [comment]
I'll leave the selection behavior like it is now, but I agree with you it's not great.

**Review by KoBeWi — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks! And congrats for your first merged Godot contribution :tada:

**Comment by hsandt** [comment]
Nice, I'll have to  try this one!

While I agree that deleting selected entries should clear selection in general, I see a point, on single delete, in selecting the new sprite replacing the deleted sprite at the same slot (or the one before if deleting the last sprite), so user can chain delete if they want to. This would have to be combined with https://github.com/godotengine/godot-proposals/issues/9387 so user keeps focus and can chain usage of the Delete key, or do some other operations like changing the Frame Duration multiplier (it really shines when doing many of these operations - like "merging" identical sprite frames together by removing all the duplicates and setting the Frame Duration multiplier).

Currently, clicking the bin button actually does that, so you can chain delete by clicking on the bin button multiple times. Now it's a technique I learned *because batch selection was not implemented* and it becomes much less useful now that we can delete multiple sprite frames at once. However it's worth considering the current behavior of single delete, whether with Delete key or bin button, to decide whether or not to maintain selection.

That said, in the case of multi-selection delete, I agree it's confusing. In fact, in an app like Ubuntu's Image Viewer / Eye of Gnome, when the Image Gallery is opened at the bottom, if you multi-select images and Delete them, the selection will fallback back to the image right after the last selected deleted entry (or before if last entry was deleted). So still a unique selection.

To sum up, we may want to preserve the single delete behavior of falling back to the next available entry, but when deleting multiple frames, it may be less confusing to fall back selection to just the next available entry after the last selected frame (whether "last" means "last selected" or "last in chronological order" would remain to be decided).
